File: Dynamic_Programming/1458. Max Dot Product of Two Subsequences.cpp
/*
1458. Max Dot Product of Two Subsequences
*/

// TC:O(n^2) SC:O(n^2)
class Solution {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size();
        int n = nums2.size();
        nums1.insert(nums1.begin(), -1);
        nums2.insert(nums2.begin(), -1);

        vector<vector<int>> dp(m+1, vector<int>(n+1,INT_MIN/2));
        dp[0][0] = 0;
        int res = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                dp[i][j] = max(nums1[i]*nums2[j],dp[i-1][j-1] + nums1[i]*nums2[j]);
                dp[i][j] = max(dp[i][j], max(dp[i-1][j], dp[i][j-1]));
            }
        }
        return dp[m][n];
    }
};

/*
給你兩個數組從裡面挑出一樣長的subsequence 兩兩相乘在加總後最大值為多少
雙序列型DP
dp[i][j]...
dp[i][j] = max(dp[i][j], )
i -> j 兩個都選 or 兩個都不選 dp[i-1][j-1] or 一個不選 dp[i-1][j], dp[i][j-1]
這題另外要注意負數情況 如果dp[i-1][j-1]反而拖累我 直接選nums1[i]*nums2[j]
*/

File: Dynamic_Programming/1458. Max Dot Product of Two Subsequences.py
# TC:O(n^2) SC:O(n^2) -> O(n)
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        # 選 or 不選
        m,n = len(nums1), len(nums2)
        dp = [float('-inf')] * (n+1)
        dp[0] = 0
        for i, x in enumerate(nums1):
            prev = dp[0]
            dp[0] = float('-inf')
            for j, y in enumerate(nums2):
                tmp = dp[j+1]
                dp[j+1] = max(prev + nums1[i] * nums2[j], nums1[i] * nums2[j], dp[j], dp[j+1])
                prev = tmp
        return dp[n]
        """
        遞推 二維
        dp = [[float('-inf')] * (n+1) for _ in range(m+1)]
        dp[0][0] = 0
        for i, x in enumerate(nums1):
            for j, y in enumerate(nums2):
                dp[i+1][j+1] = max(dp[i][j] + nums1[i] * nums2[j], nums1[i] * nums2[j], dp[i+1][j], dp[i][j+1])
        return dp[m][n]
        """
        """
        遞歸 
        @cache
        def dfs(i, j) -> int:
            if i < 0 or j < 0:
                return  float('-inf')
            return max(dfs(i-1, j-1) + nums1[i] * nums2[j], dfs(i-1,j), dfs(i,j-1),nums1[i] * nums2[j])
        return dfs(m-1, n-1)
        """
